### PR TITLE
watched_tasks: maintain a list of spawned async tasks and propagate errors

### PR DESCRIPTION
**Pull request by hnez**
This PR is based on an idea from [a discussion](https://github.com/linux-automation/tacd/pull/43#pullrequestreview-1650219550) in PR #43, where it was noticed that errors from tasks and threads propagate nowhere.

This PR fixed that by assembling a list of threads and tasks that should __not__ complete before the end of the program and propagating all errors from these threads and tasks to `main()` and beyond.

This touches a lot of places, as we spawn quite a lot of async_std tasks (the `tacd` logs now say "Spawned 58 tasks tasks and 5 threads").

This PR allows us to add better error handling for tasks and threads (we can for example use the `?` operator to propagate errors out and do not have to resort to `.unwrap()` to at least crash the `tacd` from within a task on error). Adding this better error handling is out of scope for this PR, as it is already large enough.

### CODE REVIEW

**Comment by hnez** [comment]
The failed `cargo deny` run should be fixed by the dependency updates in #42 and has nothing to do with this PR.

File: src/adc/iio/hardware.rs
Comment (KarlK90) [inline src/adc/iio/hardware.rs:365]: Can you put this into the expect call? In the future we want to deny unwraps completely.

File: src/adc/iio/hardware.rs
Comment (KarlK90) [inline src/adc/iio/hardware.rs:388]: Can you put this into the expect call? In the future we want to deny unwraps completely.

File: src/adc/iio/hardware.rs
Comment (KarlK90) [inline src/adc/iio/hardware.rs:418]: Can you put this into the expect call? In the future we want to deny unwraps completely.

**Review by KarlK90 — CHANGES_REQUESTED** [review]
Nice! I like the changes and implementation. I thought whether or not the async runtime provides mechanisms to query running tasks and see if anything failed like [`tokio-console`](https://docs.rs/tokio-console/latest/tokio_console/), but this might be messy (or not even possible). Anyway, there are some minor things that should be corrected but apart from that LGTM.

File: src/watched_tasks.rs
Comment (hnez) [reply src/watched_tasks.rs:113]: We previously used `async_std:task::spawn()` which did [`.expect()`](https://docs.rs/async-std/latest/src/async_std/task/spawn.rs.html#31) in the background.
But I will have a look …

File: src/watched_tasks.rs
Comment (KarlK90) [inline src/watched_tasks.rs:79]:
```suggestion
        let handle = self
            .handle
            .take()
            .with_context(|| "Failed to get thread handle")?;
```

File: src/watched_tasks.rs
Comment (KarlK90) [reply src/watched_tasks.rs:113]: Sure, but as we implement our own api now it would be good to be consistent for both spawning `tasks` and `threads`.

Comment (hnez) [reply src/watched_tasks.rs]: I've implemented this in a separate commit so it can be reviewed separately, as it touches a whole lot of files.

File: src/adc/iio/hardware.rs
Comment (hnez) [reply src/adc/iio/hardware.rs]: Sure. I hope to get rid of these panics rather sooner than later when #48 is merged, but until then we can have `.expect()` here.

File: src/adc/iio/hardware.rs
Comment (hnez) [reply src/adc/iio/hardware.rs]: Sure. I hope to get rid of these panics rather sooner than later when #48 is merged, but until then we can have `.expect()` here.

File: src/adc/iio/hardware.rs
Comment (hnez) [reply src/adc/iio/hardware.rs]: Sure. I hope to get rid of these panics rather sooner than later when #48 is merged, but until then we can have `.expect()` here.

File: src/watched_tasks.rs
Comment (hnez) [reply src/watched_tasks.rs]: Wouldn't work here as the `Future:poll()` implementation has to return a `Poll<>`, but I came up with another solution:

https://github.com/linux-automation/tacd/blob/69d52f700eb62493607edb6786486a6796b2d13a/src/watched_tasks.rs#L99-L108

File: src/adc/iio/hardware.rs
Comment (KarlK90) [reply src/adc/iio/hardware.rs]: Nice, lgtm!

File: src/adc/iio/hardware.rs
Comment (KarlK90) [reply src/adc/iio/hardware.rs]: Nice, lgtm!

File: src/adc/iio/hardware.rs
Comment (KarlK90) [reply src/adc/iio/hardware.rs]: Nice, lgtm!

File: src/watched_tasks.rs
Comment (KarlK90) [reply src/watched_tasks.rs]: Thanks, lgtm.

**Comment by hnez** [comment]
While addressing your comments I have also changed the way `ThreadHandle` notifies a task about the completion of a thread.
Using a `async_std::channel` there was somewhat of a hack so I do not have to work with `Future` and `Waker` directly, because that sounded somewhat scary.
As it turns out working with `Waker`s is not scary at all, so we can just use one here:

https://github.com/linux-automation/tacd/blob/69d52f700eb62493607edb6786486a6796b2d13a/src/watched_tasks.rs#L29-L54

This does however mean that you have to re-review that part.

**Comment by hnez** [comment]
The failed `cargo clippy` run should be fixed by #49 and has nothing to do with this PR.

File: src/watched_tasks.rs
Comment (KarlK90) [reply src/watched_tasks.rs]: Yes, I've seen this change. I'm not that familiar with async Rust (yet) but from my understanding it seems sound.

**Review by KarlK90 — APPROVED** [review]
This looks good to me now and will definitely improve the error handling story in the code base. Feel free to merge.

**Review by KarlK90 — APPROVED** [review]

**Comment by jluebbe** [comment]
Adding a custom wrapper around standard APIs seems very invasive to me. Is this really such an unusual problem that there is no idiomatic way?

**Comment by hnez** [comment]
Hi @jluebbe

> no idiomatic way

Not in `async_std` itself, so we would have to use a crate for it. I had a look around and found [`async_nursery`](https://docs.rs/async_nursery/) which may be coerced into doing what we we want (minus the thread handling), but I would prefer not to use it because it does not look too mature (and does not support threads in addition to tasks).

I quite like the tailor-made solution implemented in this PR and think 206 not-that-dense lines of code are an acceptable complexity compared to adding a new dependency to handle this for us.

I would prefer merging this as-is rather than adding a new dependency.

**Comment by hnez** [comment]
The recently merged pull requests #33 and #34 made some changes to this PR necessary, as both of them spawn tasks.
Please re-review affected files.

**Review by KarlK90 — APPROVED** [review]

**Comment by hnez** [comment]
I've rebased this PR on top of recent `main` and think it should be the next one we merge.

**Comment by KarlK90** [comment]
I remember that @jluebbe wanted some written commentary why we choose to implement a bespoke solution and not use `async_nursery` or tokio's [JoinSet](https://docs.rs/tokio/1.33.0/tokio/task/struct.JoinSet.html). Could you add this to `watched_tasks.rs`? After that I'm happy to approve this.

**Comment by hnez** [comment]
I've added a comment …

https://github.com/linux-automation/tacd/blob/6c7a2cd7fb57b626fb3f8abeac0bf367493eeb81/src/watched_tasks.rs#L11-L29

… that talks about why we went with our own task spawning wrapper instead of using something off-the-shelf.

**Review by KarlK90 — APPROVED** [review]
@hnez The comment summarizes the discussion very well. You have my approval - feel free to merge.

**Pull request merged**
